### PR TITLE
Typehints

### PR DESCRIPTION
**Pull request by gmponos**
Basically I tried to typehint most of the functions.

### CODE REVIEW

File: src/Middleware/RoutingMiddleware.php [hunk 1]
```diff
@@ -11,13 +13,13 @@
...
 {
-    public function __invoke($request, $response, $next)
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, $next)
     {
         try {
             $params = (array)$request->getAttribute('params', []);
             if (empty($params['controller'])) {
                 $path = $request->getUri()->getPath();
-                $request = $request->withAttribute('params', Router::parse($path, $request->getMethod()));
+                $request = $request->withAttribute('params', Router::parse($path));
```
Comment (markstory) [inline src/Middleware/RoutingMiddleware.php]: This line should not be changed. In `3.next` `Router::parse()`supports 2 arguments.

File: src/Runner.php [hunk 2]
```diff
@@ -11,15 +11,24 @@
...
 {
     protected $index;
+
+    /**
+     * @var MiddlewareStack
+     */
     protected $middleware;
 
-    public function run($middleware, ServerRequestInterface $request, ResponseInterface $response)
+    public function run(MiddlewareStack $middleware, ServerRequestInterface $request, ResponseInterface $response)
```
Comment (markstory) [inline src/Runner.php:27]: I'm no longer sure it is a good idea to type hint on concrete implementations. It makes swapping implementations out harder than no typehint does.

Comment (gmponos) [inline src/Runner.php:27]: To tell you the truth about this line I was not sure either for the same reason. But some lines below on the invoke function of the same class you are using the get method from the middleware which implies that the property `$middleware` is an object and must have the `get` function.

File: src/Middleware/RoutingMiddleware.php (hunk 1 above)
Comment (gmponos) [inline src/Middleware/RoutingMiddleware.php]: :D, Didn't knew that. I used version 3.2.5. I will put it back the way it was.

File: src/Runner.php (hunk 2 above)
Comment (markstory) [inline src/Runner.php:27]: Yes there is a required set of methods. However, typehinting against a concrete implementation requires that any changes be done as subclasses. If instead the typehint was against an interface someone could extend the library without inheritance.

File: src/Middleware/AssetMiddleware.php [hunk 3]
```diff
@@ -154,7 +156,13 @@ protected function deliverAsset($request, $response, $file)
...
     }
 
-    protected function getType($file)
+    /**
+     * Return the type from a File object
+     *
+     * @param File $file The file from which you get the type
+     * @return string
+     */
+    protected function getType(File $file)
```
Comment (markstory) [inline src/Middleware/AssetMiddleware.php:165]: I'm not convinced that typehints on protected methods are ever that valuable.

File: src/Server.php [hunk 4]
```diff
@@ -87,9 +92,9 @@ public function setApp(BaseApplication $app)
     /**
      * Get the current application.
      *
-     * @return Spekkoek\BaseApplication The application that will be run.
+     * @return BaseApplication The application that will be run.
      */
-    public function getApp(BaseApplication $app)
```
Comment (markstory) [inline src/Server.php:92]: Thanks :+1:

**Comment by markstory** [comment]
Thanks! I removed the typehints that referenced concrete implementations as I'm not convinced they are a good idea. Sure they give earlier runtime warnings, but I don't feel that typehints on concrete types make userland extension/re-implementation easier/better.

**Comment by gmponos** [comment]
Basically I also wanted to close the MR.
I thought about it and ended up that it might mess with what you have in your mind for any future implementation.

Anyway I hope I helped in a few spots..

**Comment by markstory** [comment]
@gmponos The help is appreciated. I would rather take it slow with adding typehints as removing them becomes a backwards incompatible change. Adding them does too though :smile:

**Pull request merged**